org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -0,0 +1,1170 @@
+            return;
+        }
+
+        try {
+            ss_socket = new ServerSocket(debug_port);
+            ss_socket.setReuseAddress(true);
+            ss_socket.setReceiveBufferSize(BUFFER_SIZE);
+
+            c_socket = new Socket(ss_socket.getInetAddress(), ss_socket.getLocalPort());
+            c_socket.setReuseAddress(true);
+            c_socket.setReceiveBufferSize(BUFFER_SIZE);
+            c_socket.setSendBufferSize(BUFFER_SIZE);
+
+            s_socket = ss_socket.accept();
+            s_socket.setReuseAddress(true);
+            s_socket.setReceiveBufferSize(BUFFER_SIZE);
+            s_socket.setSendBufferSize(BUFFER_SIZE);
+
+            s_ostream = s_socket.getOutputStream();
+
+            c_ostream = c_socket.getOutputStream();
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to enable debug socket logging! " + e.getMessage(), e);
+        }
+    }
+
+    public void beginHandshake() {
+        debug("JSSEngine: beginHandshake()");
+
+        // We assume beginHandshake(...) is the entry point for initializing
+        // the buffer. In particular, wrap(...) / unwrap(...) *MUST* call
+        // beginHandshake(...) if ssl_fd == null.
+
+        // ssl_fd == null <-> we've not initialized anything yet.
+        if (ssl_fd == null) {
+            init();
+        }
+
+        // Always, reset the handshake status, using the existing
+        // socket and configuration (which might've been just created).
+        if (SSL.ResetHandshake(ssl_fd, as_server) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to begin handshake: " + errorText(PR.GetError()));
+        }
+    }
+
+    public void closeInbound() {
+        debug("JSSEngine: closeInbound()");
+
+        PR.Shutdown(ssl_fd, PR.SHUTDOWN_RCV);
+        is_inbound_closed = true;
+    }
+
+    public void closeOutbound() {
+        debug("JSSEngine: closeOutbound()");
+
+        PR.Shutdown(ssl_fd, PR.SHUTDOWN_SEND);
+        is_outbound_closed = true;
+    }
+
+    public String getHostname() {
+        return hostname;
+    }
+
+    public Runnable getDelegatedTask() {
+        debug("JSSEngine: getDelegatedTask()");
+
+        // We fake being a non-blocking SSLEngine. In particular, we never
+        // export tasks as delegated tasks (e.g., OCSP checking), so this
+        // method will always return null.
+
+        return null;
+    }
+
+    public SSLEngineResult.HandshakeStatus getHandshakeStatus() {
+        debug("JSSEngine: getHandshakeStatus()");
+        try {
+            updateHandshakeState();
+        } catch (SSLException se) {
+            // Do nothing -- updateHandshakeStatus() should throw the same
+            // exception again, so it is safe to discard it. We also can't
+            // just pass the SSLException to the caller of this method, as
+            // they aren't expecting it.
+        }
+
+        return handshake_state;
+    }
+
+    public SecurityStatusResult getStatus() {
```

Comment:
`null` gets returned in the following conditions:

 1. `ssl_fd` is `null`, so the result of this function is also `null`.
 2. The security status function fails for some reason.

The latter only occurs when the socket is somehow bad (it shouldn't be -- we call this function when trying to handshake so we would've learned a lot earlier).

So that leaves the first case. We call `getStatus()` when trying to handshake, which assumes `ssl_fd` is not `null` -- so we would've crashed earlier. So it can return NULL and I think that's just fine. :-)